internals/overlord/state/identities.go

```diff
@@ -47,28 +52,63 @@ type LocalIdentity struct {
 	UserID uint32
 }
 
+// BasicIdentity holds identity configuration specific to the "basic" type
+// (for HTTP basic authentication).
+type BasicIdentity struct {
+	Password string // Note: this is the sha512-crypt hashed password.
+}
+
+// This is used to ensure we send a well-formed identity Name.
+var identityNameRegexp = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9_\-]*$`)
+
 // validate checks that the identity is valid, returning an error if not.
-func (d *Identity) validate() error {
+func (d *Identity) validate(name string) error {
+	if d == nil {
+		return errors.New("identity must not be nil")
+	}
+
+	if !identityNameRegexp.MatchString(name) {
+		return fmt.Errorf("identity name %q invalid: must start with an alphabetic character and only contain alphanumeric characters, underscore, and hyphen", d.Name)
+	}
+
+	return d.validateExcludingName()
+}
+
+// validateExcludingName checks that the identity is valid, returning an error if not.
+func (d *Identity) validateExcludingName() error {
 	if d == nil {
 		return errors.New("identity must not be nil")
 	}
 
 	switch d.Access {
-	case AdminAccess, ReadAccess, UntrustedAccess:
+	case AdminAccess, ReadAccess, MetricsAccess, UntrustedAccess:
+		// Check basic type access level.
+		if d.Basic != nil && d.Access != MetricsAccess {
+			return fmt.Errorf("basic identity can only have %q access, got %q", MetricsAccess, d.Access)
+		}
 	case "":
-		return fmt.Errorf("access value must be specified (%q, %q, or %q)",
-			AdminAccess, ReadAccess, UntrustedAccess)
+		return fmt.Errorf("access value must be specified (%q, %q, %q, or %q)",
+			AdminAccess, ReadAccess, MetricsAccess, UntrustedAccess)
 	default:
-		return fmt.Errorf("invalid access value %q, must be %q, %q, or %q",
-			d.Access, AdminAccess, ReadAccess, UntrustedAccess)
+		return fmt.Errorf("invalid access value %q, must be %q, %q, %q, or %q",
+			d.Access, AdminAccess, ReadAccess, MetricsAccess, UntrustedAccess)
 	}
 
-	switch {
-	case d.Local != nil:
-		return nil
-	default:
-		return errors.New(`identity must have at least one type ("local")`)
+	gotType := false
+	if d.Local != nil {
+		gotType = true
+	}
+	if d.Basic != nil {
```

Comment:
Is it worth validating that `Basic` and `Local` are mutually exclusive? It sounds like a logic error, but then again, we can't prevent all the user's creative ways of breaking things

Reply:
An identity may have multiple types, this is by design, see the [Pebble identities spec](https://docs.google.com/document/d/1nASgUt-piV94i1cpFsbEPRk_xFbZt8AskgL97gh1Dgg/edit?tab=t.0) for more details.

Reply:
Ah I see, thanks for the explanation!